src/main/java/io/axoniq/axonserver/connector/ResultStreamPublisher.java

```diff
@@ -0,0 +1,145 @@
+    }
+
+    private class ResultStreamSubscription implements Subscription {
+
+        private final Subscriber<? super Message> subscriber;
+        private final ResultStream<Message> resultStream;
+        private final AtomicLong requested = new AtomicLong(0);
+        private final AtomicBoolean signalGate = new AtomicBoolean(false);
+        private final AtomicBoolean cancelled = new AtomicBoolean(false);
+        private final AtomicBoolean subscribed = new AtomicBoolean(false);
+        private final AtomicBoolean completed = new AtomicBoolean(false);
+
+        private ResultStreamSubscription(Subscriber<? super Message> subscriber,
+                                         ResultStream<Message> resultStream) {
+            this.subscriber = subscriber;
+            this.resultStream = resultStream;
+            this.resultStream.onAvailable(this::signal);
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                subscriber.onError(new IllegalArgumentException("negative subscription request"));
+                return;
+            }
+            requested.updateAndGet(current -> current + Math.min(Long.MAX_VALUE - current, n));
+            signal();
+        }
+
+        @Override
+        public void cancel() {
+            logger.debug("The call has been cancelled.");
+            cancelled.set(true);
+            resultStream.close();
+        }
+
+        private void signal() {
+            while (canSignal() && signalGate.compareAndSet(false, true)) {
+                try {
+
+                    long requests = requested.get();
+                    long counter = 0;
+                    for (int i = 0; i < requests; i++) {
+                        if (!resultStream.isClosed() && resultStream.peek() != null) {
+                            subscriber.onNext(resultStream.next());
+                            counter--;
+                        } else {
+                            break;
+                        }
+                    }
+                    requested.getAndAccumulate(counter, Long::sum);
+
+                    if (resultStream.isClosed()) {
+                        subscriber.onComplete();
+                        completed.set(true);
+                    }
+                    resultStream.getError().ifPresent(this::onError);
+                } catch (InterruptedException e) {
+                    signalGate.set(false);
+                    Thread.currentThread().interrupt();
+                } catch (Exception ex) {
+                    this.onError(ex);
+                } finally {
+                    signalGate.set(false);
+                }
+            }
+        }
+
+        private void onError(Throwable error) {
+            logger.debug("An error occurred accessing the ResultStream.", error);
+            subscriber.onError(error);
+            completed.set(true);
+        }
+
+
+        private boolean canSignal() {
```

Comment:
Since the "signal" is already given when this method is executed, I'd recommend renaming this to something that indicates a check for work. Maybe "canConsume"?

Reply:
Done